File: 134-Gas Station-加油站/134-Gas Station-加油站.js
/**
 * @param {number[]} gas
 * @param {number[]} cost
 * @return {number}
 */
 var canCompleteCircuit = function(gas, cost) {
    let min = Infinity
    let lossSum = 0

    for(let i = 0; i < gas.length; i++){
        let loss = gas[i] - cost[i]
        lossSum += loss
        min = min < lossSum ? min : lossSum
    }

    if(lossSum < 0) return -1;
    if(min >= 0) return 0

    for(let i = gas.length - 1; i >= 0; i--){
        let loss = gas[i] - cost[i]
        min += loss
        if(min >= 0){
            return i
        }
    }
    return -1
};

File: 134-Gas Station-加油站/134-Gas Station-加油站.java
class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int start = 0;
        int curSum = 0;
        int totalSum = 0;

        for(int i = 0; i < gas.length; i++){
            curSum += gas[i] - cost[i];
            totalSum += gas[i] - cost[i];

            if(curSum < 0){
                start = (i + 1) % gas.length;
                curSum = 0;
            }
        }
        if(totalSum < 0) return -1;
        return start;
    }
}